python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+            a value of ``.fits.gz``.
+        """
+        extensions = self._pathLib(self.path).suffixes
+        return "".join(extensions)
+
+    def join(self, path: str) -> ButlerURI:
+        """Create a new `ButlerURI` with additional path components including
+        a file.
+
+        Parameters
+        ----------
+        path : `str`
+            Additional file components to append to the current URI. Assumed
+            to include a file at the end. Will be quoted depending on the
+            associated URI scheme.
+
+        Returns
+        -------
+        new : `ButlerURI`
+            New URI with any file at the end replaced with the new path
+            components.
+
+        Notes
+        -----
+        Schemeless URIs assume local path separator but all other URIs assume
+        POSIX separator if the supplied path has directory structure. It
+        may be this never becomes a problem but datastore templates assume
+        POSIX separator is being used.
+        """
+        new = self.dirname()  # By definition a directory URI
+
+        # new should be asked about quoting, not self, since dirname can
+        # change the URI scheme for schemeless -> file
+        if new.quotePaths:
+            path = urllib.parse.quote(path)
+
+        newpath = self._pathModule.normpath(self._pathModule.join(new.path, path))
+        new._uri = new._uri._replace(path=newpath)
+        # Declare the new URI not be dirLike unless path ended in /
+        if not path.endswith(self._pathModule.sep):
+            new.dirLike = False
+        return new
+
+    def relative_to(self, other: ButlerURI) -> Optional[str]:
+        """Return the relative path from this URI to the other URI.
+
+        Parameters
+        ----------
+        other : `ButlerURI`
+            URI to use to calculate the relative path. Must be a parent
+            of this URI.
+
+        Returns
+        -------
+        subpath : `str`
+            The sub path of this URI relative to the supplied other URI.
+            Returns `None` if there is no parent child relationship.
+            Scheme and netloc must match.
+        """
+        if self.scheme != other.scheme or self.netloc != other.netloc:
+            return None
+
+        enclosed_path = self._pathLib(self.relativeToPathRoot)
+        parent_path = other.relativeToPathRoot
+        subpath: Optional[str]
+        try:
+            subpath = str(enclosed_path.relative_to(parent_path))
+        except ValueError:
+            subpath = None
+        else:
+            subpath = urllib.parse.unquote(subpath)
+        return subpath
+
+    def exists(self) -> bool:
+        """Indicate that the resource is available.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the resource exists.
+        """
+        raise NotImplementedError()
+
+    def remove(self) -> None:
+        """Remove the resource."""
+        raise NotImplementedError()
```

Comment:
I did have abstractmethod to begin with but `__new__` completely messed that up because mypy never thinks you ever get a subclass back from the `ButlerURI` factory so it would get very upset if a the base class uses one of those abstract methods (because it would rightly say that that couldn't possibly work because the base class is an abc).

Reply:
Oh, yes. Hm, right, I can't see a way to work around that, so this is probably best. That's a frustrating limitation. Probably worth noting in a comment somewhere so that no one else goes through the same experiment and finds the same failure?